File: fraiseql-python/src/fraiseql/validators.py
"""Validation engine for custom GraphQL scalars."""

from typing import Any

from fraiseql.registry import SchemaRegistry
from fraiseql.scalars import CustomScalar


class ScalarValidationError(Exception):
    """Raised when custom scalar validation fails."""

    def __init__(self, scalar_name: str, context: str, message: str) -> None:
        self.scalar_name = scalar_name
        self.context = context
        super().__init__(
            f"Scalar {scalar_name!r} validation failed in {context}: {message}"
        )


def validate_custom_scalar(
    scalar_class: type[CustomScalar],
    value: Any,
    context: str = "parse_value",
) -> Any:
    """Execute validation for a custom scalar.

    Args:
        scalar_class: The CustomScalar subclass to validate with
        value: The value to validate
        context: One of "serialize", "parse_value", or "parse_literal"

    Returns:
        The validated/converted value

    Raises:
        ScalarValidationError: If validation fails
        ValueError: If context is unknown

    Examples:
        >>> from fraiseql.validators import validate_custom_scalar
        >>> from myapp.scalars import Email
        >>>
        >>> # Parse a variable value from GraphQL
        >>> email_value = validate_custom_scalar(Email, "user@example.com")
        >>> # Returns "user@example.com"
        >>>
        >>> # Validation fails
        >>> try:
        ...     email_value = validate_custom_scalar(Email, "invalid-email")
        ... except ScalarValidationError as e:
        ...     print(f"Validation error: {e}")
    """
    scalar_name = scalar_class.name

    try:
        scalar = scalar_class()

        if context == "serialize":
            return scalar.serialize(value)
        if context == "parse_value":
            return scalar.parse_value(value)
        if context == "parse_literal":
            return scalar.parse_literal(value)

        raise ValueError(f"Unknown validation context: {context!r}")

    except ScalarValidationError:
        raise
    except ValueError as e:
        raise ScalarValidationError(scalar_name, context, str(e)) from e
    except Exception as e:
        raise ScalarValidationError(
            scalar_name, context, f"{type(e).__name__}: {str(e)}"
        ) from e


def get_all_custom_scalars() -> dict[str, type[CustomScalar]]:
    """Get all registered custom scalars.

    Returns:
        Dictionary mapping scalar names to CustomScalar classes

    Examples:
        >>> from fraiseql.validators import get_all_custom_scalars
        >>> scalars = get_all_custom_scalars()
        >>> # {'Email': <class Email>, 'Phone': <class Phone>, ...}
    """
    return SchemaRegistry.get_custom_scalars()

File: sdks/official/fraiseql-typescript/src/validators.ts
/**
 * Validation engine for custom GraphQL scalars.
 */

import { SchemaRegistry } from "./registry";
import { CustomScalar } from "./scalars";

/**
 * Raised when custom scalar validation fails.
 */
export class ScalarValidationError extends Error {
  /**
   * Creates a new ScalarValidationError.
   *
   * @param scalarName - Name of the scalar that failed validation
   * @param context - The validation context ("serialize", "parseValue", or "parseLiteral")
   * @param message - The underlying error message
   */
  constructor(
    public scalarName: string,
    public context: "serialize" | "parseValue" | "parseLiteral",
    message: string
  ) {
    super(
      `Scalar ${JSON.stringify(scalarName)} validation failed in ${context}: ${message}`
    );
    this.name = "ScalarValidationError";
  }
}

/**
 * Execute validation for a custom scalar.
 *
 * @param scalarClass - The CustomScalar subclass to validate with
 * @param value - The value to validate
 * @param context - One of "serialize", "parseValue", or "parseLiteral"
 * @returns The validated/converted value
 * @throws ScalarValidationError if validation fails
 * @throws Error if context is unknown
 *
 * @example
 * ```typescript
 * import { validateCustomScalar } from "fraiseql/validators"
 * import { Email } from "./scalars"
 *
 * // Parse a variable value from GraphQL
 * const emailValue = validateCustomScalar(Email, "user@example.com", "parseValue")
 * // Returns "user@example.com"
 *
 * // Validation fails
 * try {
 *   validateCustomScalar(Email, "invalid-email", "parseValue")
 * } catch (e) {
 *   // ScalarValidationError: Scalar "Email" validation failed in parseValue: Invalid email
 * }
 * ```
 */
export function validateCustomScalar(
  scalarClass: typeof CustomScalar,
  value: unknown,
  context: "serialize" | "parseValue" | "parseLiteral" = "parseValue"
): unknown {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any -- instantiating abstract class subclass
  const instance = new (scalarClass as new () => any)();
  const scalarName = instance.name;

  try {
    switch (context) {
      case "serialize":
        return instance.serialize(value);
      case "parseValue":
        return instance.parseValue(value);
      case "parseLiteral":
        return instance.parseLiteral(value);
      default:
        throw new Error(`Unknown validation context: ${context}`);
    }
  } catch (error) {
    if (error instanceof ScalarValidationError) {
      throw error;
    }

    const message =
      error instanceof Error ? error.message : String(error);
    throw new ScalarValidationError(scalarName, context, message);
  }
}

/**
 * Get all registered custom scalars.
 *
 * @returns Map of scalar names to CustomScalar classes
 *
 * @example
 * ```typescript
 * import { getAllCustomScalars } from "fraiseql/validators"
 *
 * const scalars = getAllCustomScalars()
 * // Map { "Email" => class Email, "Phone" => class Phone, ... }
 * ```
 */
export function getAllCustomScalars(): Map<string, typeof CustomScalar> {
  return SchemaRegistry.getCustomScalars();
}
